src/main/java/mybatis/mapper/GioHangMapper.java

```diff
@@ -11,57 +11,90 @@
 
-  //cập nhật số lượng sản phẩm trong giỏ hàng, khi người dùng bấm nút thêm giỏ hàng nhiều lần
+  // cập nhật số lượng sản phẩm trong giỏ hàng, khi người dùng bấm nút thêm giỏ
+  // hàng nhiều lần
   final String INCREASE_SL_SP = "UPDATE `gio_hang` SET `ma_khach_hang` = 1, `so_luong` = `so_luong` + 1 " +
-  " WHERE `gio_hang`.`ma_khach_hang` = #{maKhachHang} AND `gio_hang`.`ma_san_pham` = #{maSanPham}";
+      " WHERE `gio_hang`.`ma_khach_hang` = #{maKhachHang} AND `gio_hang`.`ma_san_pham` = #{maSanPham}";
+
   @Update(INCREASE_SL_SP)
-  public int increaseSoLuongSP(@Param("maKhachHang") int maKhachHang,@Param("maSanPham") String maSanPham);
+  public int increaseSoLuongSP(@Param("maKhachHang") int maKhachHang, @Param("maSanPham") String maSanPham);
 
-    //cập nhật số lượng sản phẩm trong giỏ hàng
-    final String UPDATE_SO_LUONG_SP = "UPDATE `gio_hang` SET `ma_khach_hang` = #{maKhachHang}, `ma_san_pham` = #{maSanPham}, " +
-    "`so_luong` = #{soLuong} " +
-    "WHERE `gio_hang`.`ma_khach_hang` = #{maKhachHang} AND `gio_hang`.`ma_san_pham` = #{maSanPham};";
-    @Update(UPDATE_SO_LUONG_SP)
-    public int updateSoLuongSP_In_GioHang(@Param("maKhachHang") int maKhachHang,@Param("maSanPham") String maSanPham,@Param("soLuong") int soLuong);
+  // cập nhật số lượng sản phẩm trong -- trang giỏ hàng ---
+  final String UPDATE_SO_LUONG_SP = "UPDATE `gio_hang` SET `ma_khach_hang` = #{maKhachHang}, `ma_san_pham` = #{maSanPham}, "
+      +
+      "`so_luong` = #{soLuong} " +
+      "WHERE `gio_hang`.`ma_khach_hang` = #{maKhachHang} AND `gio_hang`.`ma_san_pham` = #{maSanPham};";
 
+  @Update(UPDATE_SO_LUONG_SP)
+  public int updateSoLuongSP_In_GioHang(
+      @Param("maKhachHang") int maKhachHang,
+      @Param("maSanPham") String maSanPham,
+      @Param("soLuong") int soLuong
+    );
 
-  //Xóa sản phẩm ra khỏi giỏ hàng
+  // Xóa sản phẩm ra khỏi giỏ hàng
   final String DELETE_SP_GIO_HANG = "DELETE FROM `gio_hang` WHERE `gio_hang`.`ma_khach_hang` = #{maKhachHang} " +
-  "AND `gio_hang`.`ma_san_pham` = #{maSanPham}";
+      "AND `gio_hang`.`ma_san_pham` = #{maSanPham}";
+
   @Delete(DELETE_SP_GIO_HANG)
-  public int deleteSP(@Param("maKhachHang") int maKhachHang,@Param("maSanPham") String maSanPham);
+  public int deleteSP(@Param("maKhachHang") int maKhachHang, @Param("maSanPham") String maSanPham);
+
+  /*
+   * Trước khi cập nhật trong trang giỏ hàng, hoặc tăng số lượng khi click chuột
+   * đặt hàng nhiều lần
+   * Kiểm tra xem số lượng của giỏ hàng có lớn 0 và vượt quá `so_luong` trong
+   * table `san_pham` hay không?
+   */
+  final String CHECK_SP_HET_HANG = "select sp.so_luong as 'so_luong_hien_co',sp.ten_san_pham,gh.so_luong as 'so_luong_can_mua' " +
+    "from gio_hang gh join san_pham sp on sp.ma_san_pham = gh.ma_san_pham " +
+    "where gh.ma_khach_hang = #{maKhachHang} AND gh.so_luong = #{soLuong} and sp.ma_san_pham = #{maSanPham} AND gh.so_luong > sp.so_luong;";
```

Comment:
câu này phải là 
where gh.ma_khach_hang = #{maKhachHang} AND and gh.ma_san_pham = #{maSanPham} AND #{soLuong} > sp.so_luong;
mới đúng chứ

Reply:
ờ ha

Reply:
còn chỗ nào lỗi ko ô?

Reply:
ông test thử lại chưa, cái này là tui đoán thôi

Reply:
Test thử lại được ô